### PR TITLE
BIP39 cosmetics

### PR DESCRIPTION
**Pull request by semenov-vladyslav**
# Description of change

Please write a summary of your changes and why you made them.

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/keys/bip39.rs [hunk 1]
```diff
@@ -36,33 +38,33 @@ pub enum Error {
 }
 
 /// Reference to a normalized (unicode NFKD) mnemonic.
-#[derive(Clone, Copy)]
-pub struct MnemonicRef<'a>(&'a str);
+pub struct MnemonicRef(str);
```
Comment (DaughterOfMars) [inline src/keys/bip39.rs:42]: `#[repr(transparent)]`?

File: src/keys/bip39.rs [hunk 2]
```diff
@@ -36,33 +38,33 @@ pub enum Error {
...
 
-impl<'a> Deref for MnemonicRef<'a> {
+impl Deref for MnemonicRef {
     type Target = str;
     fn deref(&self) -> &str {
-        self.0
+        unsafe { core::mem::transmute(self) }
```
Comment (Alex6323) [inline src/keys/bip39.rs:48]: I think in the Rust ecosystem it's good practice to accompany an `unsafe` with a  `//Safety: <why is it still safe>` doc.

File: src/keys/bip39.rs [hunk 3]
```diff
@@ -36,33 +38,33 @@ pub enum Error {
...
     }
 }
 
-impl<'a> TryFrom<&'a str> for MnemonicRef<'a> {
+impl ToOwned for MnemonicRef {
+    type Owned = Mnemonic;
+    fn to_owned(&self) -> Mnemonic {
+        Mnemonic(unsafe { core::mem::transmute::<_, &str>(self).to_owned() })
```
Comment (Alex6323) [inline src/keys/bip39.rs]: same here

File: src/keys/bip39.rs [hunk 4]
```diff
@@ -36,33 +38,33 @@ pub enum Error {
...
+    }
+}
+
+impl<'a> TryFrom<&'a str> for &'a MnemonicRef {
     type Error = Error;
     fn try_from(mnemonic_str: &'a str) -> Result<Self, Error> {
         if is_nfkd(mnemonic_str) {
-            Ok(MnemonicRef(mnemonic_str))
+            Ok(unsafe { core::mem::transmute(mnemonic_str) })
```
Comment (Alex6323) [inline src/keys/bip39.rs:64]: same here ... and the others down below.

File: src/keys/bip39.rs [hunk 5]
```diff
@@ -190,14 +246,27 @@ impl AsRef<[u8; 64]> for Seed {
 }
 
 impl Seed {
-    pub fn null() -> Self {
+    pub fn new() -> Self {
```
Comment (thibault-martinez) [inline src/keys/bip39.rs]: `Seed::new` may feel like it's generating a random one, at least `null` was explicit :/

File: src/keys/bip39.rs (hunk 2 above)
Comment (DaughterOfMars) [reply src/keys/bip39.rs:48]:
```rust
// Safety: I have no idea what I am doing
```

File: src/keys/bip39.rs (hunk 1 above)
Comment (Alex6323) [inline src/keys/bip39.rs:42]: `#[repr(transparent)]` ?

Comment (DaughterOfMars) [reply src/keys/bip39.rs:42]: This is needed for all types you are transmuting btw. Otherwise it really is unsafe.

File: src/keys/bip39.rs (hunk 5 above)
Comment (Alex6323) [reply src/keys/bip39.rs]: I agree, why changing it back? Wasn't `null()` just being introduced?

File: src/keys/bip39.rs (hunk 1 above)
Comment (semenov-vladyslav) [reply src/keys/bip39.rs:42]: Exactly. Done.

File: src/keys/bip39.rs (hunk 2 above)
Comment (semenov-vladyslav) [reply src/keys/bip39.rs:48]: Added comment.

File: src/keys/bip39.rs (hunk 3 above)
Comment (semenov-vladyslav) [reply src/keys/bip39.rs]: Done

File: src/keys/bip39.rs [hunk 6]
```diff
@@ -158,15 +193,37 @@ impl<'a> TryFrom<&'a str> for PassphraseRef<'a> {
 #[derive(Clone, Zeroize, ZeroizeOnDrop)]
 pub struct Passphrase(String);
 
-impl From<String> for Passphrase {
-    fn from(unnormalized_passphrase: String) -> Self {
-        Self(unnormalized_passphrase.chars().nfkd().collect())
+impl Passphrase {
+    pub fn new() -> Self {
```
Comment (thibault-martinez) [inline src/keys/bip39.rs:197]: Does it really make sense to provide this?

File: src/keys/bip39.rs [hunk 7]
```diff
@@ -195,9 +259,22 @@ impl Seed {
     }
 }
 
+impl Default for Seed {
+    fn default() -> Self {
```
Comment (thibault-martinez) [inline src/keys/bip39.rs]: Not sure how to feel about this, people may use it 🤷🏻‍♂️

Comment (DaughterOfMars) [reply src/keys/bip39.rs]: Is it invalid to use a null seed? And if so, shouldn't the `null` fn be private?

Comment (thibault-martinez) [reply src/keys/bip39.rs]: It's not technically invalid, but providing a Seed that everyone can use feels wrong

**Review by thibault-martinez — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

**Pull request merged**